TASVideos/Services/ModuleParamHelpers.cs

```diff
@@ -0,0 +1,95 @@
+			.Assembly
+			.GetTypes()
+			.Where(t => t.BaseType != null
+				&& t.BaseType.IsGenericType
+				&& t.BaseType.GetGenericTypeDefinition() == typeof(ModuleParameterTypeAdapter<>))
+			.ToDictionary(
+				t => t.BaseType!.GetGenericArguments()[0],
+				t => (IModuleParameterTypeAdapter)Activator.CreateInstance(t)!);
+
+		public static readonly IDictionary<string, Type> ViewComponents = Assembly
+			.GetAssembly(typeof(WikiModuleAttribute))
+			!.GetTypes()
+			.Where(t => t.GetCustomAttribute(typeof(WikiModuleAttribute)) != null)
+			.ToDictionary(tkey => ((WikiModuleAttribute)tkey.GetCustomAttribute(typeof(WikiModuleAttribute))!).Name, tvalue => tvalue, StringComparer.InvariantCultureIgnoreCase);
+
+		// TODO: reuse code above
+		public static readonly IDictionary<string, Type> TextComponents = Assembly
+			.GetAssembly(typeof(WikiModuleAttribute))
+			!.GetTypes()
+			.Where(t => t.GetCustomAttribute(typeof(WikiModuleAttribute)) != null)
+			.Where(t => t.GetCustomAttribute(typeof(TextModuleAttribute)) != null)
+			.ToDictionary(tkey => ((WikiModuleAttribute)tkey.GetCustomAttribute(typeof(WikiModuleAttribute))!).Name, tvalue => tvalue, StringComparer.InvariantCultureIgnoreCase);
+
+		public static IDictionary<string, object?> GetParameterData(
+			TextWriter w,
+			string name,
+			MethodInfo invokeMethod,
+			WikiPage pageData,
+			IReadOnlyDictionary<string, string> pp)
+		{
+			var paramObject = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase)
+			{
+				{ "pageData", pageData }
+			};
+
+			var paramCandidates = invokeMethod
+				.GetParameters()
+				.Where(p => !paramObject.ContainsKey(p.Name!)); // filter out any already supplied parameters
+
+			foreach (var paramCandidate in paramCandidates)
+			{
+				var paramType = paramCandidate.ParameterType;
+				var adapterKeyType = paramType;
+				var doNullableWrap = paramType.IsValueType
+					&& (!paramType.IsGenericType || paramType.GetGenericTypeDefinition() != typeof(Nullable<>));
+
+				if (doNullableWrap)
+				{
+					adapterKeyType = typeof(Nullable<>).MakeGenericType(adapterKeyType);
+				}
+
+				if (!ParamTypeAdapters.TryGetValue(adapterKeyType, out var adapter))
+				{
+					// These should all exist at compile time.
+					throw new InvalidOperationException($"Unknown ViewComponent Argument Type: {adapterKeyType}");
+				}
+
+				pp.TryGetValue(paramCandidate.Name!, out var ppValue);
+				var result = adapter.Convert(ppValue);
+
+				if (result == null)
+				{
+					// Conversion failed.  See if the parameter type is a failable type.
+					var needsNonNull = paramType.IsValueType && doNullableWrap
+						|| !paramType.IsValueType && paramType.ToContextualType().Nullability == Nullability.NotNullable;
+					if (needsNonNull)
+					{
+						// TODO: Better styling, or something
+						w.Write($"MODULE ERROR for `{name}`: Missing parameter value for {paramCandidate.Name}");
```

Comment:
With the way this has been refactored, this method shouldn't receive the `TextWriter` and this isn't how it should report that failure.  We should return some failure state to the caller which can then handle output.